### ISSUE TITLE
Walking Around Corners

### ISSUE THREAD

**Issue by Morphan1** [CitizensDev/Citizens2#1730]
When player NPCs reach a barrier that they are meant to walk around, they sometimes get too close to the walls and prevent themselves from properly walking around. This has effects ranging anywhere from sliding sideways along a wall until they forcibly pass it to the more extreme cases of slamming into a corner and never continuing, as shown in the screenshot below.

(The full path taken is in black, the remaining desired path is in red.)
![image](https://user-images.githubusercontent.com/4834997/55362924-7d4e4b00-54a9-11e9-98fc-49e146db2e91.png)

You *can* use `/npc move` further along in the desired path to force it to continue until it hits the next corner, so it does seem to be trying its best, but it will not move again on its own.

Non-player NPCs seem relatively unaffected by this, and can navigate the maze without issue.

A likely irrelevant side note: this issue exists both with `use-new-finder` enabled and disabled. In the case of it being disabled, I was unable to generate a path through the miniature maze in the screenshot altogether.

**Comment by mcmonkey4eva** [comment]
This is a fairly problematic issue.
In the process of testing #1740 I had to give the NPC an extra door because it kept slamming into the wall next to the door instead of walking through the open door.

**Comment by fullwall** [comment]
Try the latest build with the new pathfinder - should be better. For doors, the code is
   ```
 if (door.isOpen()) {
                BlockFace targetFace = door.getFacing().getOppositeFace();
                destX = vector.getX() + targetFace.getModX();
                destZ = vector.getZ() + targetFace.getModZ();
            }
``` 
(the open door corner)
I wonder if it should be set somewhere else?

**Comment by mcmonkey4eva** [comment]
Not fixed. Maybe a bit better?

Testing with versions: `git-Spigot-766dab2-b9a2a74 (MC: 1.14)` and `Citizens version 2.0.25-SNAPSHOT (build 1654)`

**For the doorway test:** NPCs seem to be able to slide past the doorway now... but they still need to slowly slide past (as opposed to knowing how to just walk through the open space).
**For the two points test:** The NPC now consistently stops a significant margin before reaching the destination. It appears to be closer to the correct distance margin at least though. (looks to be greater than 1 block length, with distance-margin: 1. Note that new-finder false with a zombie mob leaves the zombie exactly on the destination block, rather than offset at all).
**For the maze-like tight corners test:** The NPC is sliding pretty harshly along the sides but is at least making it through. However, new-finder false actually seems to be able to handle it now - actually better than new-finder true does.

Image of rebuilt testing environment, if relevant:
![testland](https://i.alexgoodwin.media/i/misc/a12510.png)

At the top are the two points to walk between, towards the bottom is the minimaze that I've asked the NPC to go all the way through. Separating these two areas is a long wall containing the testing doorway.

**Comment by fullwall** [comment]
For two points test you'd expect stopping about 1 block away since it checks for distanceSquared <= distanceMargin.

**Comment by mcmonkey4eva** [comment]
At distance margin 0.1 (which, helpfully, actually works now) it gets roughly this close:
![almost](https://i.alexgoodwin.media/i/misc/2218b3.png)
Its destination is at the center of the stone block. When approaching from the opposite direction it's the same amount away (it looks to be one full block _plus_ the distance margin)

**Comment by mcmonkey4eva** [comment]
Verified that all corner handling trouble with new-finder true is fixed on `git-Spigot-e4be9b0-5e0b63c (MC: 1.14)` and `Citizens version 2.0.25-SNAPSHOT (build 1664)`

Additionally by setting `/npc pathopt --path-distance-margin 0.5 --distance-margin 0.5` (need to specify both at once at the moment as https://github.com/CitizensDev/Citizens2/commit/09226946040aefd940398fe71c532d719ecfe626#diff-65cd2d1148da260738c5a38b5367ac53R1134 is bad, due to non-matching flag names) the NPC will step fully onto the destination block, which is good enough.
A path distance of `0.1` gets the NPC exactly on its goal location, however since it duplicate the distance margin currently it will work a bit odd (the NPC moves its head side to side while walking as the distance margin is too low)

Once that typo in the command is fixed, I believe all issues listed here are fully resolved.

**Comment by mcmonkey4eva** [comment]
with the new commit, all points of this issue are resolved
